# Exploratory Data Analysis - Crime Data Los Angeles

This notebook is used to explore the questions for the Data Scientist Exercise in July 2020.

- How many crimes were reported over the past 5 years?
- List the top five reported crimes for each year for the past 5 years.
- What are the most common MO codes? Have these changed over the past 5 years?
- How else has reported crime changed over time in the City?
- Based on your analysis, please share any changes to services, programs, or policies that the City should consider.

##  Load libraries

In [10]:
# Import libraries
import os, sys, subprocess
import json
import pandas as pd
import numpy as np
from langdetect import detect
import pickle

pd.set_option('display.max_colwidth', -1)

pd.set_option('display.max_rows', 1000)
#pd.set_option('display.max_columns', 500)

In [14]:
# load project config
terminal_call = ! git rev-parse --show-toplevel
repo_path=terminal_call[0]
project_config_path = os.path.join(repo_path,'project_config.json')

with open(project_config_path,'r') as fp: 
    project_config = json.load(fp)

In [25]:
# import custom module to look at trends
module_path = os.path.join(repo_path,project_config['project_module_relative_path'])
sys.path.append(module_path)

import trends
from trends import get_top_trends as gt

from importlib import reload

# Load in data

In [20]:
# load in data that was collected
df = pickle.load(open("../data/crime_data.pkl", "rb"))

## How many crimes were reported over the past 5 years?

In [21]:
# Check how many crimes were reported over the past 5 years
# Calculate current date and offset by 5 years
(df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))).sum()

1117864

### A: 1,117,864 crimes were reported in the past 5 years

## List the top five reported crimes for each year for the past 5 years.

In [142]:
reload(get_top_trends)

<module 'trends.get_top_trends' from '/Users/vivianpeng/git/LA_innovation_team/src/trends/get_top_trends.py'>

In [55]:
# Select the last 5 years as a dataframe
df_5 = df[df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))]

In [143]:
# Find the top 5 reported crimes for each year
# Crime is listed in the data under 'crm_cd'
# Custom function 'top_trends' groups by column and selects the top 'n' 
gt.top_trends(df_5, column='year', variable='crm_cd', n=5)

year  crm_cd  count
0    2015  624     8853 
1    2015  510     8573 
2    2015  440     7866 
3    2015  330     7316 
4    2015  354     7297 
131  2016  510     18353
132  2016  624     17942
133  2016  330     16779
134  2016  440     14814
135  2016  310     14558
261  2017  624     19074
262  2017  510     18757
263  2017  330     18082
264  2017  310     15279
265  2017  440     14774
397  2018  624     19427
398  2018  330     18070
399  2018  510     17129
400  2018  440     15420
401  2018  310     14814
532  2019  624     19008
533  2019  330     16750
534  2019  440     15455
535  2019  510     15427
536  2019  740     12859
668  2020  510     9760 
669  2020  624     8198 
670  2020  330     6893 
671  2020  310     6695 
672  2020  740     6368

In [81]:
test

crm_cd
year crm_cd        
2015 624     8853  
     510     8573  
     440     7866  
     330     7316  
     354     7297  
2016 510     18353 
     624     17942 
     330     16779 
     440     14814 
     310     14558 
2017 624     19074 
     510     18757 
     330     18082 
     310     15279 
     440     14774 
2018 624     19427 
     330     18070 
     510     17129 
     440     15420 
     310     14814 
2019 624     19008 
     330     16750 
     440     15455 
     510     15427 
     740     12859 
2020 510     9760  
     624     8198  
     330     6893  
     310     6695  
     740     6368

In [51]:
test.dtypes

crm_cd    int64
dtype: object

In [50]:
test["crm_cd"].apply(crime_codes.get)

year  crm_cd
2015  624       None
      510       None
      440       None
      330       None
      354       None
2016  510       None
      624       None
      330       None
      440       None
      310       None
2017  624       None
      510       None
      330       None
      310       None
      440       None
2018  624       None
      330       None
      510       None
      440       None
      310       None
2019  624       None
      330       None
      440       None
      510       None
      740       None
2020  510       None
      624       None
      330       None
      310       None
      740       None
Name: crm_cd, dtype: object

## What are the most common MO codes? Have these changed over the past 5 years?

In [36]:
# Find the top 10 most common MO codes across all the years (2010-2020)
df_5["mocodes"].value_counts(ascending = False).head(10)

0344         93842
0329         44407
1501         18472
0325         12737
0416         12317
0329 1300    8302 
1822         8234 
0344 1300    6090 
0344 1606    5898 
0329 1307    4808 
Name: mocodes, dtype: int64

In [37]:
# Find the top 10 MO codes for each year
# MO code is listed in the data under 'mocodes'
# Custom function 'top_trends' groups by column and selects the top 'n' 
gt.top_trends(df_5, column='year', variable='mocodes', n=5)

year  mocodes  
2015  0344         12619
      0329         4784 
      1501         2013 
      0325         1632 
      0416         1573 
2016  0344         22719
      0329         9713 
      1501         4215 
      0325         3037 
      0416         2939 
2017  0344         20205
      0329         9811 
      1501         4436 
      0416         2938 
      0325         2737 
2018  0344         17485
      0329         9032 
      1501         3705 
      0416         2375 
      0325         2354 
2019  0344         15849
      0329         8156 
      1501         2984 
      0325         2327 
      0416         1977 
2020  0344         4965 
      0329         2911 
      1501         1119 
      0344 1300    706  
      0329 1300    698  
Name: mocodes, dtype: int64